xla/service/gpu/runtime/command_buffer_cmd.cc

```diff
@@ -1988,4 +1991,252 @@ CommandBufferCmd::BufferUsageVector CollectiveBroadcastCmd::buffers() {
   return buffer_usage;
 }
 
+//===----------------------------------------------------------------------===//
+// DynamicSliceFusionCmd
+//===----------------------------------------------------------------------===//
+
+DynamicSliceFusionCmd::DynamicSliceFusionCmd(
+    ExecutionStreamId execution_stream_id,
+    std::unique_ptr<CommandBufferCmdSequence> embedded_commands,
+    std::vector<std::optional<BufferAllocation::Slice>> arguments,
+    std::vector<std::unique_ptr<BufferAllocation>> fake_allocations,
+    std::vector<std::optional<std::vector<DynamicSliceThunk::Offset>>> offsets,
+    std::vector<std::optional<Shape>> orig_shapes,
+    std::vector<std::optional<Shape>> sliced_shapes,
+    std::vector<std::optional<uint64_t>> offset_byte_sizes)
+    : CommandBufferCmd(CommandBufferCmdType::kDynamicSliceFusionCmd,
+                       execution_stream_id),
+      embedded_commands_(std::move(embedded_commands)),
+      fake_allocations_(std::move(fake_allocations)) {
+  // Zip all arguments together to create a list of SliceDef.
+  for (auto [arg, offsets, orig_shape, sliced_shape, offset_byte_size] :
+       llvm::zip_equal(arguments, offsets, orig_shapes, sliced_shapes,
+                       offset_byte_sizes)) {
+    slices_.push_back(DynamicSliceThunk::SliceDef{
+        std::move(arg),
+        std::move(offsets),
+        std::move(orig_shape),
+        std::move(sliced_shape),
+        std::move(offset_byte_size),
+    });
+  }
+
+  for (auto [argument_idx, slice] : llvm::enumerate(slices_)) {
+    embeded_to_origin_slice_map_[argument_idx] = slice.embedded_thunk_argument;
+  }
+
+  // Find how many offsets we might have to transfer from device to host and
+  // pre-compute host allocation requirements.
+  for (DynamicSliceThunk::SliceDef& slice : slices_) {
+    offsets_allocs_base_.push_back(offsets_allocs_size_);
+    if (slice.sliced_shape.has_value()) {
+      offsets_allocs_size_ += slice.sliced_shape->rank() * sizeof(int64_t);
+    }
+  }
+}
+
+bool DynamicSliceFusionCmd::force_update() {
```

Comment:
can you add documentation on how decition is made, it's quire hard to parse it from the codr